src/dawn/SIR/SIR.cpp

```diff
@@ -50,6 +193,259 @@ static bool pointerMapComparison(const sir::GlobalVariableMap& map1,
+      }
+    }
+    return std::make_pair(output, true);
+  }
+}
+
+} // anonymous namespace
+
+std::pair<std::string, bool> SIR::comparison(const SIR& rhs) const {
+  std::string output;
+  if((Stencils.size() != rhs.Stencils.size()))
+    return std::make_pair("[SIR missmatch] number of stencils do not match\n", false);
+  if(StencilFunctions.size() != rhs.StencilFunctions.size())
+    return std::make_pair("[SIR missmatch] number of stencil functions does not match\n", false);
+  if(GlobalVariableMap.get()->size() != rhs.GlobalVariableMap.get()->size())
+    return std::make_pair("[SIR missmatch] number of global variables does not match\n", false);
+
+  if(!Stencils.empty() &&
+     !std::equal(Stencils.begin(), Stencils.end(), rhs.Stencils.begin(),
+                 pointeeComparison<sir::Stencil>)) {
+    output += "[SIR missmatch] stencils do not match\n";
+    for(unsigned i = 0; i < Stencils.size(); ++i) {
+      auto tmp = pointeeComparisonWithOutput(Stencils[i], rhs.Stencils[i]);
+      if(tmp.second == false) {
+        output += tmp.first;
+      }
+    }
+    return std::make_pair(output, false);
+  }
+  if(!StencilFunctions.empty() &&
+     !std::equal(StencilFunctions.begin(), StencilFunctions.end(), rhs.StencilFunctions.begin(),
+                 pointeeComparison<sir::StencilFunction>)) {
+    output += "[SIR missmatch] stencil functions do not match\n";
+    for(unsigned i = 0; i < StencilFunctions.size(); ++i) {
+      auto tmp = pointeeComparisonWithOutput(StencilFunctions[i], rhs.StencilFunctions[i]);
+      if(tmp.second == false) {
+        output += "Missmatch of function " + StencilFunctions[i]->Name + "\n" + tmp.first;
+      }
+    }
+    return std::make_pair(output, false);
+  }
+  if(!GlobalVariableMap.get()->empty() &&
+     !pointerMapComparison(*(GlobalVariableMap.get()), *(rhs.GlobalVariableMap.get()))) {
+    auto a = pointerMapComparisonWithOutput(*GlobalVariableMap.get(), *rhs.GlobalVariableMap.get());
+    return std::make_pair(a.first, false);
+  }
+
+  return std::make_pair("", true);
+}
+
+std::pair<std::string, bool> sir::Stencil::comparison(const sir::Stencil& rhs) const {
+  std::string output;
+  if(Fields.size() != rhs.Fields.size()) {
+    output += "[Stencil missmatch] number of Fields does not match\n";
+    return std::make_pair(output, false);
+  }
+  if(Name != rhs.Name) {
+    output += "[Stencil missmatch] Stencil names do not match\n";
+    return std::make_pair(output, false);
+  }
+  if(!(Attributes == rhs.Attributes)) {
+    output += "[Stencil missmatch] Stencil attibutes do not match\n";
+    output += dawn::format("of Stencil %s and %s", Name, rhs.Name);
+    return std::make_pair(output, false);
+  }
+  if(!StencilDescAst->getRoot().get()->equals(rhs.StencilDescAst->getRoot().get())) {
+    output += "[Stencil missmatch] Stencil ASTs Do not match\n";
+    auto a = compareAst(StencilDescAst, rhs.StencilDescAst);
+    if(a.second)
+      dawn_unreachable("Stencils should not match");
```

Comment:
Better in this case use `DAWN_ASSERT_MSG(a.second, "stencils should not match");`. Unreachables are more a compiler hint in switch statements to indicate impossible cases.